### ISSUE TITLE
Error: read ECONNRESET     at TCP.onStreamRead (internal/stream_base_commons.js:111:27)

### ISSUE THREAD

**Issue by kamalzaman** [Joystream/storage-node-joystream#48]
I have successfully running a storage node. Until today I found it exited with this error below:
```
Staked, proceeding.
________                     _____
______(_)__________  __________  /__________________ _______ ___
_____  /_  __ \_  / / /_  ___/  __/_  ___/  _ \  __ `/_  __ `__ \
____  / / /_/ /  /_/ /_(__  )/ /_ _  /   /  __/ /_/ /_  / / / / /
___  /  \____/_\__, / /____/ \__/ /_/    \___/\__,_/ /_/ /_/ /_/
/___/         /____/
API server started; API docs at http://localhost:3000/swagger.json
DHT started on port 3060
DHT JSON-RPC service started on port 3090
Sync server started at { address: '::', family: 'IPv6', port: 3030 }
events.js:174
      throw er; // Unhandled 'error' event
      ^

Error: read ECONNRESET
    at TCP.onStreamRead (internal/stream_base_commons.js:111:27)
Emitted 'error' event at:
    at Socket.onerror (_stream_readable.js:713:12)
    at Socket.emit (events.js:189:13)
    at emitErrorNT (internal/streams/destroy.js:82:8)
    at emitErrorAndCloseNT (internal/streams/destroy.js:50:3)
    at process._tickCallback (internal/process/next_tick.js:63:19)

```

I am running Ubuntu 16.04. Port 3000, 3030, 3060, 3090 already opened

**Comment by bwhm** [comment]
Hey @kamalzaman 

1. You were successful in running it after #43 right? Did you stop it yourself, or did it crash?

The fact that it starts with  `Staked, proceeding.` leads me to believe you have done _something_ between starting/stopping (that you shouldn't 😄  )

2. Do you mind pasting the what happened since it shut down, and what commands you've done in between?
3. Is your joystream-node running (I doubt you'd get that far otherwise, but still).

**Comment by kamalzaman** [comment]
Hi @bwhm 

1. If I remember correctly, I kill the process once, because I need it to run with screen/tmux.  :D

2. After I kill the process, I run it again with `js_storage --key-file 5CbhqfnkbQPFwVV1tBopLaQFZmnijBC6ouofEVzHsPKk4qtH.json`
and no error afterward until it exited today.

3. My joystream-node still running and syncing

**Comment by bwhm** [comment]
I see I'm getting `Staked, proceeding` myself when restarting, so that wasn't it. Starting with `screen` works well for me (on mac though).

This is grasping at straws, but:
1. `js_storage --help` and find + open the config file. Paste the content (scramble if you want to hide something)
2. `js_storage list` and check if the repository output matches `/path/to/storage/repos` (if you just used `js_storage create` the first time, it will be in the `storage/repos`

More likely:
I think [this](https://stackoverflow.com/questions/54037384/nodejs-error-read-econnreset-at-tcp-onstreamread-internal-stream-base-common) implies that we need to change something, and that the problem seems to occur on VMs for some reason...

**Comment by kamalzaman** [comment]
So actually restarting will get slashed or not?

Anyway here is the log

1.

```
{
        "port": 3000,
        "syncPort": 3030,
        "syncPeriod": 30000,
        "dhtPort": 3060,
        "dhtRpcPort": 3090,
        "storage": "/home/ubuntu/storage-node-joystream/storage",
        "storageType": "hyperdrive",
        "keyFile": "5CbhqfnkbQPFwVV1tBopLaQFZmnijBC6ouofEVzHsPKk4qtH.json"
}
```

2. `js_storage list` output `169af99b-c6a2-5d52-9607-b046826183da`
matched with path `/home/ubuntu/storage-node-joystream/storage/repos/16/9a# ls169af99b-c6a2-5d52-9607-b046826183da
`

I also would like to add that I am using AWS instance for storage node

**Comment by bwhm** [comment]
> So actually restarting will get slashed or not?

Not yet. 😄 You are actually the only one that has gotten paid (in joy)!
> I also would like to add that I am using AWS instance for storage node

That is what we expect in the future, so it can't be an issue long term. Ours is of course also hosted somewhere...

Unfortunately, this is above my pay grade to solve. Let me see if I can delete and get it restored without having to setup new key(s).

**Comment by kamalzaman** [comment]
Ah, I was thinking the definition of getting slashed is blacklisted/banned forever, where it is actually just getting kicked out from validator/storage provider for one era. Is that correct?

Actually, I found that I am still an actor on storage role even though my storage node is offline. Does it mean I am still getting reward (joy)?
Ok then, I will run my storage node again, and see if I met this error again

**Comment by bwhm** [comment]
First. Do you mind checking the size and content in the storage folder? 

> Ah, I was thinking the definition of getting slashed is blacklisted/banned forever, where it is actually just getting kicked out from validator/storage provider for one era. Is that correct?

I don't think we are autobanning/slashing yet.

> Actually, I found that I am still an actor on storage role even though my storage node is offline. Does it mean I am still getting reward (joy)?

As I have not gotten any rewards (I have started/restarted many times). I'm quite sure it checks if we are running.

---

I was able to delete the folder and restore. 

It might work just as well to just `git pull` (in storage folder) and `yarn && yarn build`, but I didn't try that. If you don't mind testing that first, it would be cool!

If not, before you delete, make sure to copy these out first:
1. The key(s) 
(You only need really need `5CbhqfnkbQPFwVV1tBopLaQFZmnijBC6ouofEVzHsPKk4qtH.json` for that)
2. The `storage` folder. The whole thing.

3. Then, delete the `joystream-storage-node` folder
4. clone to same location, `yarn run build` , and `npm install -g`
5. move the key(s) and `storage` folder back into to it, and start with 
`js_storage --key-file 5CbhqfnkbQPFwVV1tBopLaQFZmnijBC6ouofEVzHsPKk4qtH.json`


---

Also out of curiosity, how did you kill it in the first place? ctrl+c?

**Comment by kamalzaman** [comment]
Well @bwhm it seems like my storage node didn't sync,
```

root@ip-172-31-8-203:/home/ubuntu/storage-node-joystream/storage# du -ah
4.0K    ./repos/16/9a/169af99b-c6a2-5d52-9607-b046826183da/content/key
4.0K    ./repos/16/9a/169af99b-c6a2-5d52-9607-b046826183da/content/bitfield
4.0K    ./repos/16/9a/169af99b-c6a2-5d52-9607-b046826183da/content/signatures
4.0K    ./repos/16/9a/169af99b-c6a2-5d52-9607-b046826183da/content/tree
20K     ./repos/16/9a/169af99b-c6a2-5d52-9607-b046826183da/content
4.0K    ./repos/16/9a/169af99b-c6a2-5d52-9607-b046826183da/metadata/key
4.0K    ./repos/16/9a/169af99b-c6a2-5d52-9607-b046826183da/metadata/secret_key
4.0K    ./repos/16/9a/169af99b-c6a2-5d52-9607-b046826183da/metadata/data
4.0K    ./repos/16/9a/169af99b-c6a2-5d52-9607-b046826183da/metadata/bitfield
4.0K    ./repos/16/9a/169af99b-c6a2-5d52-9607-b046826183da/metadata/signatures
4.0K    ./repos/16/9a/169af99b-c6a2-5d52-9607-b046826183da/metadata/tree
28K     ./repos/16/9a/169af99b-c6a2-5d52-9607-b046826183da/metadata
52K     ./repos/16/9a/169af99b-c6a2-5d52-9607-b046826183da
56K     ./repos/16/9a
60K     ./repos/16
64K     ./repos
4.0K    ./id
4.0K    ./keys
76K     .

``` 

Or is that normal?

> Also out of curiosity, how did you kill it in the first place? ctrl+c?

Yes, with simple `ctrl+c`.

> It might work just as well to just git pull (in storage folder) and yarn && yarn build, but I didn't try that. If you don't mind testing that first, it would be cool!

I will try testing `git pull` later

**Comment by kamalzaman** [comment]
Well @bwhm it seems like my storage node didn't sync,
```

/home/ubuntu/storage-node-joystream/storage# du -ah
4.0K    ./repos/16/9a/169af99b-c6a2-5d52-9607-b046826183da/content/key
4.0K    ./repos/16/9a/169af99b-c6a2-5d52-9607-b046826183da/content/bitfield
4.0K    ./repos/16/9a/169af99b-c6a2-5d52-9607-b046826183da/content/signatures
4.0K    ./repos/16/9a/169af99b-c6a2-5d52-9607-b046826183da/content/tree
20K     ./repos/16/9a/169af99b-c6a2-5d52-9607-b046826183da/content
4.0K    ./repos/16/9a/169af99b-c6a2-5d52-9607-b046826183da/metadata/key
4.0K    ./repos/16/9a/169af99b-c6a2-5d52-9607-b046826183da/metadata/secret_key
4.0K    ./repos/16/9a/169af99b-c6a2-5d52-9607-b046826183da/metadata/data
4.0K    ./repos/16/9a/169af99b-c6a2-5d52-9607-b046826183da/metadata/bitfield
4.0K    ./repos/16/9a/169af99b-c6a2-5d52-9607-b046826183da/metadata/signatures
4.0K    ./repos/16/9a/169af99b-c6a2-5d52-9607-b046826183da/metadata/tree
28K     ./repos/16/9a/169af99b-c6a2-5d52-9607-b046826183da/metadata
52K     ./repos/16/9a/169af99b-c6a2-5d52-9607-b046826183da
56K     ./repos/16/9a
60K     ./repos/16
64K     ./repos
4.0K    ./id
4.0K    ./keys
76K     .

``` 

Or is that normal?

> Also out of curiosity, how did you kill it in the first place? ctrl+c?

Yes, with simple `ctrl+c`.

> It might work just as well to just git pull (in storage folder) and yarn && yarn build, but I didn't try that. If you don't mind testing that first, it would be cool!

I will try testing `git pull` later

**Comment by bwhm** [comment]
>Well, it seems like my storage node didn't sync,
Or is that normal?

yes, I believe it's mentioned twice in the guide 😄 

> I will try testing git pull later

You would need to try that first...
Let me know either way! Feel free to post monero address here if you are comfortable. If not I have your joy address so you can prove identity later (although that is less convenient tbh)

**Comment by kamalzaman** [comment]
> Again, note that the nodes are currently not syncing, so you will have to update the node later.
Dang, I really forgot this. :satisfied:

> Let me know either way! Feel free to post monero address here if you are comfortable. If not I have your joy address so you can prove identity later (although that is less convenient tbh)
You can use my joy address, I have set my xmr address memo in it (if that is OK). 
Thanks!!

**Comment by jfinkhaeuser** [comment]
Is there anything more to that backtrace, or just what you posted?

**Comment by kamalzaman** [comment]
> Is there anything more to that backtrace, or just what you posted?

Yes only that. But I am ready to send any log if needed. 
And also I have run `git pull` yesterday. Until today my storage node still running fine.

**Comment by kamalzaman** [comment]
Got this error again, but not really sure when it is occurred
```
Staked, proceeding.
________                     _____
______(_)__________  __________  /__________________ _______ ___
_____  /_  __ \_  / / /_  ___/  __/_  ___/  _ \  __ `/_  __ `__ \
____  / / /_/ /  /_/ /_(__  )/ /_ _  /   /  __/ /_/ /_  / / / / /
___  /  \____/_\__, / /____/ \__/ /_/    \___/\__,_/ /_/ /_/ /_/
/___/         /____/
API server started; API docs at http://localhost:3000/swagger.json
DHT started on port 3060
DHT JSON-RPC service started on port 3090
Sync server started at { address: '::', family: 'IPv6', port: 3030 }
events.js:174
      throw er; // Unhandled 'error' event
      ^

Error: read ECONNRESET
    at TCP.onStreamRead (internal/stream_base_commons.js:111:27)
Emitted 'error' event at:
    at Socket.onerror (_stream_readable.js:713:12)
    at Socket.emit (events.js:189:13)
    at emitErrorNT (internal/streams/destroy.js:82:8)
    at emitErrorAndCloseNT (internal/streams/destroy.js:50:3)
    at process._tickCallback (internal/process/next_tick.js:63:19)

```

**Comment by jfinkhaeuser** [comment]
Thanks! If you set the `DEBUG='joystream*` environment variable, logs would be a lot more verbose. But I think I know where this is happening. So if you do have the debug logs, that'd be helpful, but if not I think I'll have something in a few days that will take care of this.

**Comment by kamalzaman** [comment]
> Thanks! If you set the `DEBUG='joystream*` environment variable, logs would be a lot more verbose. But I think I know where this is happening. So if you do have the debug logs, that'd be helpful, but if not I think I'll have something in a few days that will take care of this.

I'd like to help, if you still need that log. I don't know to do that though.

**Comment by jfinkhaeuser** [comment]
Start with
```
$ DEBUG='joystream*' js_storage --key-file 5CbhqfnkbQPFwVV1tBopLaQFZmnijBC6ouofEVzHsPKk4qtH.json
```

I'm not sure if I'll need it, I'll let you know, ok?

**Comment by finnyfound** [comment]
This error is occurring for me too, every time i start it and come back the next day i see this error,

```
________                     _____
______(_)__________  __________  /__________________ _______ ___
_____  /_  __ \_  / / /_  ___/  __/_  ___/  _ \  __ `/_  __ `__ \
____  / / /_/ /  /_/ /_(__  )/ /_ _  /   /  __/ /_/ /_  / / / / /
___  /  \____/_\__, / /____/ \__/ /_/    \___/\__,_/ /_/ /_/ /_/
/___/         /____/
API server started; API docs at http://localhost:3001/swagger.json
DHT started on port 3060
DHT JSON-RPC service started on port 3090
Sync server started at { address: '::', family: 'IPv6', port: 3030 }
events.js:174
      throw er; // Unhandled 'error' event
      ^

Error: read ECONNRESET
    at TCP.onStreamRead (internal/stream_base_commons.js:111:27)
Emitted 'error' event at:
    at Socket.onerror (_stream_readable.js:713:12)
    at Socket.emit (events.js:189:13)
    at emitErrorNT (internal/streams/destroy.js:82:8)
    at emitErrorAndCloseNT (internal/streams/destroy.js:50:3)
    at process._tickCallback (internal/process/next_tick.js:63:19)
```

**Comment by finnyfound** [comment]
I am running 
`DEBUG='joystream*' js_storage --key-file "myaddress.json"`
I will report back next time it crashes. Do let me know how to get the debug file though.

**Comment by bwhm** [comment]
Thanks @finnyfound!

We will make an announcement regarding the storage node later this week...

**Comment by jfinkhaeuser** [comment]
This issue will be resolved when #66 is implemented.

**Comment by mnaamani** [comment]
Since implementing synchronization of storage via IPFS this is no longer an issue. Closing
